### ISSUE TITLE
Service async Initializer called multiple times (v3)

### ISSUE THREAD

**Issue by unematiii** [L2jLiga/fastify-decorators#834]
While upgrading packages in one of my projects today, I noticed that starting from version `3.13` async initializer gets called multiple times (looks like once per injection?).

I'm wrapping `knex` as follows:

```typescript
import { FastifyInstance } from 'fastify';
import { FastifyInstanceToken, Initializer, Inject, Service } from 'fastify-decorators';
import { Knex } from 'knex';

import { Database } from '@/database/Database';

@Service()
export class DatabaseService {
  readonly connection!: Knex;

  @Inject(FastifyInstanceToken)
  readonly instance!: FastifyInstance;

  public constructor() {
    this.connection = Database.getInstance().connection;
    this.instance.addHook('onClose', this.onClose.bind(this));
  }

  /**
   * This gets called on each @Inject call?
   * x1: Service A -> DatabaseService
   * x1: Service B -> Service A -> DatabaseService
   * etc
   */
  @Initializer()
  public async init() {
    await this.connection.raw(`select current_database()`);
  }

  private async onClose() {
    await this.connection.destroy();
  }
}
```

Nothing in documentation indicates that I should keep track on initialization myself (idempotent initializer), so I'm assuming its a bug?

**Comment by akbarsahata** [comment]
getting the same issue too. so I reverted to 3.12.0

**Comment by L2jLiga** [comment]
Hi there, I think issue is comes from here:
https://github.com/L2jLiga/fastify-decorators/blob/467df70605a07ef58e34d39d733cc50ce3cd1c26/lib/decorators/helpers/inject-dependencies.ts#L53

Gonna check now

**Comment by L2jLiga** [comment]
Could you check this version? Issue should be fixes now

[fastify-decorators-3.14.1.tgz](https://github.com/L2jLiga/fastify-decorators/files/9939314/fastify-decorators-3.14.1.tgz)

**Comment by akbarsahata** [comment]
> Could you check this version? Issue should be fixes now
> 
> [fastify-decorators-3.14.1.tgz](https://github.com/L2jLiga/fastify-decorators/files/9939314/fastify-decorators-3.14.1.tgz)

@L2jLiga tried it on my development branch, it seems like this solved the problem 👍🏻

**Comment by unematiii** [comment]
Hi! Thanks for taking the time. I can confirm that initializer problems is gone. However, with `3.14.1` one of my controllers fails with cryptic error:

```
    err: {
      "type": "TypeError",
      "message": "Cannot use 'in' operator to search for 'Symbol(fastify-decorators.service-injection)' in 300",
      "stack":
          TypeError: Cannot use 'in' operator to search for 'Symbol(fastify-decorators.service-injection)' in 300
              at Object.hasServiceInjection (/Project/node_modules/fastify-decorators/decorators/helpers/null:65:1)
              at Object.get (/Project/node_modules/fastify-decorators/decorators/helpers/null:42:1)
              at Proxy.createToken (/Project/src/api/services/token.service.ts:86:25)
              at Proxy.issueToken (/Project/src/api/services/token.service.ts:54:17)
              at processTicksAndRejections (internal/process/task_queues.js:95:5)
              at Proxy.getToken (/Project/src/api/controllers/tokens/tokens.controller.ts:23:14)
    }
```
`Project/src/api/services/token.service.ts:86:25` simply reads couple of variables from class instance:
```typecript
const { a, b } = this;
```

**Comment by L2jLiga** [comment]
Hi @unematiii, 
Can you check this version please?
[fastify-decorators-3.14.1.tgz](https://github.com/L2jLiga/fastify-decorators/files/9979093/fastify-decorators-3.14.1.tgz)

**Comment by akbarsahata** [comment]
> Hi @unematiii, Can you check this version please? [fastify-decorators-3.14.1.tgz](https://github.com/L2jLiga/fastify-decorators/files/9979093/fastify-decorators-3.14.1.tgz)

getting the same problem but with version 3.12.1. when trying [that](https://github.com/L2jLiga/fastify-decorators/files/9979093/fastify-decorators-3.14.1.tgz) version, calling dependencies multiple times happened again

**Comment by L2jLiga** [comment]
Thanks for feedback! I need better tests then to ensure it's not called multiple times 🤔

**Comment by unematiii** [comment]
> Hi @unematiii, Can you check this version please? [fastify-decorators-3.14.1.tgz](https://github.com/L2jLiga/fastify-decorators/files/9979093/fastify-decorators-3.14.1.tgz)

Seems that it breaks the initializer again, however, the error I mentioned above is gone. I will try to create a small repro code to test this, if it helps.

**Comment by unematiii** [comment]
So created a repro [here](https://github.com/unematiii/fastify-decorators-test). 

```
npm run dev
```
and

```
$ curl localhost:3000
```
gives me 

```
{"statusCode":500,"error":"Internal Server Error","message":"Cannot use 'in' operator to search for 'Symbol(fastify-decorators.service-injection)' in ServiceB"}%
```

While console:
```
[INFO] 21:10:25 ts-node-dev ver. 2.0.0 (using ts-node ver. 10.9.1, typescript ver. 4.8.4)
ServiceA::init()
ServiceA::init()
ServiceA::init()
ServiceA::init()
Hello from service ServiceA
```

**Comment by L2jLiga** [comment]
Hi @unematiii, thanks for your repro! I've added code to the [examples/async-initializer](https://github.com/L2jLiga/fastify-decorators/tree/v3/examples/async-initializer).

I think this version will work perfectly fine and solves the issue: 
[fastify-decorators-3.14.1.tgz](https://github.com/L2jLiga/fastify-decorators/files/10033375/fastify-decorators-3.14.1.tgz)

@unematiii, @akbarsahata can you check please? If everything ok for you I will publish it as 3.15.0

**Comment by unematiii** [comment]
Hi, it does solve the previous issues (initializer and 'in' operator), but I spotted another one: 

```javascript
{
    "type": "TypeError",
    "message": "knex.join is not a function",
    "stack":
        TypeError: knex.join is not a function
            at (/Users/mati/Workspace/projectX/src/database/utils.ts:25:8
            at Proxy.servicBMethod ((/Users/mati/Workspace/projectX/src/api/public/services/ServiceB.service.ts:31:56)
            at Proxy.serviceAMethod ((/Users/mati/Workspace/projectX/src/api/public/services/ServiceA.service.ts:25:6)
            at Proxy.methodA ((/Users/mati/Workspace/projectX/src/api/public/controllers/controllerA.controller.ts:32:35)
            at ControllerA.methodProxy ((/Users/mati/Workspace/projectX/node_modules/fastify-decorators/decorators/helpers/null:31:1)
            at Object.<anonymous> ((/Users/mati/Workspace/projectX/node_modules/fastify-decorators/decorators/strategies/null:73:1)
            at preHandlerCallback ((/Users/mati/Workspace/projectX/node_modules/fastify/lib/handleRequest.js:128:37)
            at preValidationCallback ((/Users/mati/Workspace/projectX/node_modules/fastify/lib/handleRequest.js:112:5)
            at next ((/Users/mati/Workspace/projectX/node_modules/fastify/lib/hooks.js:168:7)
            at handleResolve ((/Users/mati/Workspace/projectX/node_modules/fastify/lib/hooks.js:185:5)
}
```

This is an error from POST to controllerA ->  (Injected) serviceA -> (Injected) serviceB -> calls method with resource (knex instance) form injected Service (ServiceC.knex_instance) with async initializer. Logging knex to console in ServiceB gives me ```[Function: bound knex]```. Method `join` is indeed missing, already before `utils.ts:25:8`.

I might find some time during the weekend to also take a look at the code in this project and try understand what is going wrong (looking this stacktrace, looks like Proxy messes st up).

**Comment by L2jLiga** [comment]
Fine, at least two issues already solved, next step to fix on more (?) and don't break what is already fixed off :D

Yeah, proxy messed stacktrace a bit, however still readable. They're solving another issue - getting Request/Response within services (#387). I think it can be somehow related to how they are done, can you share the method from serviceB, initializer and utils part that causes the issue?

**Comment by unematiii** [comment]
Hi!

So ServiceC (with async init) exposes to ServiceB a knex instance `ReturnType<typeof knex>`. Utils part is a higer-order function that helps me get correct types when joining tables:

```typescript
export const JoinedTables = <T1 extends TableSchema, T2 extends TableSchema>(
  tableAName: string,
  tableBName: string,
  foreignKey: TableColumn<T1>,
  references: TableColumn<T2>,
) => {
  const getTableAColumnName = createGetColumnName<T1>(tableAName);
  const getTableBColumnName = createGetColumnName<T2>(tableBName);

  return <R extends TableColumn<T1>, S extends TableColumn<T2>>(
    knex: Knex,
    tableAColumns: R[],
    tableBColumns: S[],
  ): Knex.QueryBuilder<T1 & T2, (Pick<T1, R> & Pick<T2, S>)[]> =>
    knex
      .join(tableBName, {
        [getTableBColumnName(references)]: getTableAColumnName(foreignKey),
      })
      .select([
        ...tableAColumns.map(getTableAColumnName),
        ...tableBColumns.map(getTableBColumnName),
      ])
      .from(tableAName);
};

// Dummy example
export const OAuthSessionsWithSubject = JoinedTablesModel<OAuthSession, Subject>(
  'oauth_sessions,
  'subjects,
  'subjecte_id',
  'id',
);
```

ServiceB:
```typescript
@Service()
class ServiceB {
  @Inject(ServiceC)
  private serviceC!: ServiceC;

  async serviceBMethod() {
    const result = await OAuthSessionsWithSubject(this.serviceC.knex, ['scopes'], ['id', 'name']).where('x', 'y').first();
    // Do st else
  }
}
```

**Comment by unematiii** [comment]
It looks the culprit in the latest version you sent is the [line](https://github.com/L2jLiga/fastify-decorators/blob/b36ed40d8d266fe2fea06fa5d48d4c448f6e88be/lib/decorators/helpers/patch-methods.ts#L44):
```
 return injectedService[prop].bind(createProxy(injectedService, request, reply));
```

I can see it being executed for `injectedService:ServiceC` and `prop:knex`, which seems to mess up knex (bind).

**Comment by L2jLiga** [comment]
Aha, and if you comment out the line, then issue gone?

**Comment by L2jLiga** [comment]
Well, I simplified this proxy code and avoided using "bind", I hope this will solve the issue or at least make stacktrace bit more clearer

[fastify-decorators-3.14.1.tgz](https://github.com/L2jLiga/fastify-decorators/files/10049801/fastify-decorators-3.14.1.tgz)

**Comment by unematiii** [comment]
@L2jLiga Last version seems to fix all issues, I didn't notice anything new. Thanks!

**Comment by L2jLiga** [comment]
Published as [v3.15.0](https://github.com/L2jLiga/fastify-decorators/releases/tag/v3.15.0)
